keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,842 @@
+such as shiftfs could allow different pods to see a volume with its own IDs but
+it is not available yet. Among the possible future kernel solutions, we can
+list:
+
+- [shiftfs: uid/gid shifting filesystem](https://lwn.net/Articles/757650/)
+- [A new API for mounting filesystems](https://lwn.net/Articles/753473/)
+- [user_namespace: introduce fsid mappings](https://lwn.net/Articles/812221/)
+
+In regard to this proposal, volumes can be divided in ephemeral and
+non-ephemeral.
+
+Ephemeral volumes are associated to a **single** pod and their lifecyle is
+dependent on that pod. These are `configMap`, `secret`, `emptyDir`,
+`downwardAPI`, etc. These kind of volumes can work with any of the three
+different modes of `userNamespaceMode` as they are not shared by different pods
+and hence all the processes accessing those volumes have the same effective user
+and group IDs. Kubelet creates the files for those volumes and it can easily set
+the file ownership too.
+
+Non-ephemeral volumes are more difficult to support since they can be persistent
+and shared by multiple pods. This proposal supports volumes with two different
+strategies:
+- The `Cluster` mode makes it easier for pods to share files using volumes when
+  those don't have access permissions for `others` because the effective user
+  and group IDs on the host are the same for all the pods.
+- The semantics of semantics of `fsGroup` are respected, if it's specified it's
+  assumed to be the correct GID in the host and a 1-to-1 mapping entry for the
+  `fsGroup` is added to the GID mappings for the pod.
+
+This KEP doesn't impose any restriction on the different volumes and
+`userNamespaceMode` combinations and leaves it to users to chose the correct
+combinations based on their specific needs. For instance, if a pod access a
+shared volume containing files and folders with permissions for `others`, it can
+run in `Pod` mode. On the other hand, a process inside a pod will not be able to
+access files with mode `0700` and owned by a user different than the effective
+user of that process in a volume that doesn't support the semantics of `fsGroup`
+(doesn't support
+[`SetVolumeOwnership`](https://github.com/kubernetes/kubernetes/blob/00da04ba23d755d02a78d5021996489ace96aa4d/pkg/volume/volume_linux.go#L42)
+that updates permissions and ownership of the files to be accesible by the
+`fsGroup` group ID). Such pods should be run in `Host` mode.
+
+#### Container Runtime Support
+
+- **Docker**:
+  Docker only supports a [single ID
+  mappings](https://docs.docker.com/engine/security/userns-remap/) shared by all
+  containers running in the host. There is not support for [multiple ID
+  mappings](https://github.com/moby/moby/issues/28593) yet. Updating dockershim
+  is out of scope of this KEP as [it's going to be
+  deprecated]((https://github.com/kubernetes/enhancements/pull/1985/)) and it
+  offers a very limited support for user namespaces.
```

Comment:
How will this impact people who have not moved off of Docker? I get that `Cluster` and `Pod` mode won't work with Docker, but `Host` mode will just work like this are today?

Reply:
Yes, host mode will work just fine and with no changes needed